### DISCUSSION TITLE
CL prompting for any CL command - Hinders me from leaving SEU for CL programming

### DISCUSSION THREAD

**Discussion by richardschoen** [codefori/vscode-ibmi#2467] [Q&A]
**Is your feature request related to a problem? Please describe.**

This is not really a problem, but a feature that holds me back from leaving SEU for editing CL programs and commands. 

First off, SEU does CL command prompting really well in a CL source member when editing. RDI does command prompting, but it's slow to prompt, so that's a no go for any real productivity at the speed I code :-)

I create a lot of CL programs and rely heavily on CL command prompting to fill in optional and mandatory parameters and do some line formatting for me.

Since SEU and RDI can retrieve and create a command prompt screen, I'm pretty sure there is probably a way to do this and build an on-the-fly CL prompt dialog in VS Code.  Or possibly just fill in all the parameter markers with defaults in VS Code editor so the rest can just be filled in. 

Ex: I tried CL prompting briefly. It does appear that DCLF kind of works for this because it filled our info like this:

<img width="653" alt="image" src="https://github.com/user-attachments/assets/40569343-3356-4c42-b130-9aa785b2e344">

Created this in the editor:

```
DCLF FILE(*LIBL/File) RCDFMT(File) OPNID(*NONE) +
ALWVARLEN(*NO) ALWNULL(*NO) ALWGRAPHIC(*NO) DCLBINFLD(*DEC)
```
I suppose another way to do it would be to do something similar to what DCLF example is doing. I'm guessing there's some prompting info stored in the Code for I settings somewhere.

However, ideally retrieving command definitions on the fly from IBMi would put VS Code on par with RDI and SEU for CL programming.

Anyway I hope this makes sense. It would remove the barrier for me continuing to use SEU for CL programming.

**Describe the solution you'd like**

When I type a CL command and press F4 (or some other prompting mechanism such as the DCLF example above), locate the existing command info and create a prompt dialog to complete the command or just default the entire command with all its parameters into the VS Code editor so I can complete the command definition without needing to switch back to SEU or RDI.

**Describe alternatives you've considered**

Perhaps create a keyword settings file where common commands can be defined by the developer/user.  Similar to the DCLF example I mentioned above. In Code for i it may actually work this way already perhaps, but I would need to know where the keywords are stored. 

Ideally retrieving command definitions on the fly from IBMi would put VS Code on par with RDI and SEU for CL programming.

**Additional context**

**Comment by richardschoen** [comment]
I did see this gets installed: 
<img width="733" alt="image" src="https://github.com/user-attachments/assets/6765f2c7-b436-4b7b-87bf-37e0ab9faa0c">

There are also a couple other CL lang plugins I plan to check out. 

<img width="664" alt="image" src="https://github.com/user-attachments/assets/cdbe8ddc-879d-4e04-9758-b3c8f8d6d8c8">

<img width="652" alt="image" src="https://github.com/user-attachments/assets/889a9a71-facf-42ea-9b98-c68466ec6950">

**Comment by richardschoen** [comment]
Ah, interesting. It does look like Control-Spacebar will show me available parms for existing commands and let me select all parts, but it doesn't fill in any of the default values.

**Comment by richardschoen** [comment]
Here's what I'm visualizing. 

Instead of this for all parms. (which is how it appears to work now once I figured out the Control-Spacebar option): 
```
QSHONI/CMDLINE() SETPKGPATH() DSPSTDOUT() LOGSTDOUT() PRTSTDOUT() DLTSTDOUT() IFSSTDOUT() IFSFILE() IFSOPT() CCSID() PRTSPLF() PRTUSRDTA() PRTTXT() PRTHOLD() PRTOUTQ() OUTFILE() MBROPT() PASEJOBNAM()
```

Something like this perhaps if I select to see all parms or maybe an "All Parms with Defaults" option:

Example how it could look just picking up all the default all the keywords:
```
QSHONI/QSHEXEC SETPKGPATH(*YES) DSPSTDOUT(*NO) LOGSTDOUT(*NO) PRTSTDOUT(*NO) DLTSTDOUT(*YES) IFSSTDOUT(*NO) IFSOPT(*REPLACE) CCSID(*SAME) PRTSPLF(QSHEXECLOG) PRTUSRDTA(*NONE) PRTTXT(*NONE) PRTHOLD(*YES) PRTOUTQ(*SAME) OUTFILE(STDOUTQSH) MBROPT(*REPLACE) PASEJOBNAM(*DEFAULT)    
```
Perhaps an option to default all the keywords with a 1 or 2 keywords per line option. "All Parms with Defaults-Keyword Per Line"
```
QSHONI/QSHEXEC SETPKGPATH(*YES) DSPSTDOUT(*NO) LOGSTDOUT(*NO)  +
               PRTSTDOUT(*NO) +
               DLTSTDOUT(*YES)  +       
               IFSSTDOUT(*NO) +   
               IFSOPT(*REPLACE) +        
               CCSID(*SAME)  + 
               PRTSPLF(QSHEXECLOG)  +    
               PRTUSRDTA(*NONE)  +
               PRTTXT(*NONE)  +
               PRTHOLD(*YES) +      
               PRTOUTQ(*SAME)  +      
               OUTFILE(STDOUTQSH)  +     
               MBROPT(*REPLACE)  +
               PASEJOBNAM(*DEFAULT)  +   
```

Or possibly a prompt screen, but I could possibly settle for an option that allows me to see/prefill all my defaults perhaps with some line formatting as above.

<img width="662" alt="image" src="https://github.com/user-attachments/assets/5e7a152d-8992-465b-98e1-42ba1da25b8d">

Hopefully this is useful info.

**Comment by richardschoen** [comment]
Looks like this issue is related: 
https://github.com/IBM/vscode-clle/issues/7

All good info related to CL editing and how it could move more of us CL types forward :-)

**Comment by worksofliam** [comment]
@richardschoen this is all great info. Thanks!

I am glad you found the content assist for parameters (and it also works for special values).

I believe we can have an option to put the default parameters in automatically. I think we have access to those default values.

Regarding the program: we have wanted a prompt for sometime. Sadly we haven't had the time to build the UI for it. Perhaps something for 2025.

Again, thanks for all the info.

**Comment by SJLennon** [comment]
@richardschoen 
There is a section in the documentation on CL: https://codefori.github.io/docs/extensions/clle/

**Comment by richardschoen** [comment]
@SJLennon Thanks. There's not any way to get default parm values currently, correct ?

**Comment by SJLennon** [comment]
@richardschoen 
I'm not a developer, just a user who has contributed to the documentation, but as far as I can see there isn't a way to highlight the default value. I believe the information comes from an API call, and perhaps it doesn't specify the default parm value. For example, here the default for AUT is *LIBCRTAUT, but it looks like the values are always sorted 
alphabetically:
![image](https://github.com/user-attachments/assets/35ca16ef-4d9d-4027-9d83-3b010f0cdd03)

It might be more useful to have the wrench items at the top of the list, especially if you have a lot or program variables.  

What's the use case for knowing the default value?

**Comment by richardschoen** [comment]
The use case for seeing the default values is pretty much the same as when you prompt a CL command in SEU or RDI. It's good to do a visual reference point on the available parms and their available default values. Sometimes you will want to specify a value for a default parm to make sure someone doesn't override your choices by changing command defaults. I've seen that happen many times over the year when customers have changed system default parm values and it caused undesirable results. Thanks for feedback.
